Keyboard accessibility: Add a "Skip link" to jump back from the inspector to the selected block

afercia opened the pull request:
This PR adds a way for keyboard and assistive technologies users to jump back to the currently selected block.

Users can already use keyboard shortcuts to navigate the editor regions, see #3084 and screen reader users can use ARIA landmarks, see #2380. These tools can be used to jump from the block being edited to the Sidebar but there's no way to jump back from the sidebar to the block being edited.

This PR adds a reusable component that mimics the "skip links" already used in core. Actually, it's a button that gets revealed just when it receives keyboard focus and, when pressed, moves focus back to the currently selected block.

Related:
- coordination with #5709 
- investigation / resolution of #5847 

For now, it's placed in the block inspector in the sidebar (in the bottom right corner of the screen) but it could be useful also in the "post settings". I could see it also reused in the header toolbar and in the publishing flow, so I guess the best option would be to consider it as a generic component. In this regard, I'm not sure where the new files should be placed. For now they're in the `editor` directory. Any feedback welcome.

Screenshot:

<img width="419" alt="screen shot 2018-03-28 at 21 31 03" src="https://user-images.githubusercontent.com/1682452/38051821-b041e032-32cf-11e8-9687-dcee506e7eff.png">

Design-wise, I've made it look like the current "skip links" in core, any feedback welcome.

When a better, customizable, shortcut for `navigateRegions` will be available, see:
`Consider a mechanism to customize shortcuts, e.g. Ctrl + backtick` #3218
there will be a way for all keyboard users to jump from the selected block to the sidebar and back from the sidebar to the selected block, thus fixing #1182

afercia committed:
Skip to selected block first pass.

afercia committed:
Use withSelect.

afercia committed:
Polish.

karmatosed commented:
> Design-wise, I've made it look like the current "skip links" in core, any feedback welcome.

Whilst I would like to change those, I think keeping it the same is sensible. Changing to me seems like we'd cause friction where none is needed.

Position wise it seems to make sense tabbing through. Is it usual to have such skips in different places on screens? It totally may be, which is why asking for insights.

karmatosed reviewed (approved):
Design wise this is just like core so giving it approval for now.

aduth commented on editor/components/skip-to-selected-block/style.scss:
```diff
@@ -0,0 +1,27 @@
+.skip-to-selected-block {
```
This class name does not adhere to the Gutenberg naming conventions:

https://github.com/WordPress/gutenberg/blob/master/docs/coding-guidelines.md#naming

It should be corrected to `.editor-skip-to-selected-block`

afercia replied:
Yep, I've used `editor-` initially then removed waiting for a feedback about the placement of the whole thing,

aduth commented on editor/components/skip-to-selected-block/style.scss:
```diff
@@ -0,0 +1,27 @@
+	}
+}
+
+.edit-post-sidebar .skip-to-selected-block:focus {
```
`edit-post` is a superset of (extends) the underlying `editor`, so there should not be awareness of that context from here. These styles, if needed, should be applied within `edit-post/components/sidebar/style.scss` (also better reflecting that they're specific to the edit-post sidebar, for someone looking for those styles).

aduth commented on editor/utils/dom.js:
```diff
@@ -16,6 +16,10 @@ export function getBlockDOMNode( uid ) {
 	return document.querySelector( '[data-block="' + uid + '"]' );
 }
 
+export function getBlockWrapperDOMNode( uid ) {
```
Can we add a JSDoc block here describing the intended behavior, arguments, and return value? That we describe this as expected to return an element with tabIndex might reveal that it could potentially benefit from a more appropriate name: `getBlockTabbableWrapper` ? I don't feel strongly; proposed name is fine too.

afercia replied:
Ah yes, forgot to do that.

aduth commented on editor/components/skip-to-selected-block/index.js:
```diff
@@ -0,0 +1,37 @@
+/**
```
Aside: If "Skip To" buttons are expected to be commonplace, could make for a good base `components` component.

afercia replied:
Do you mean moving it to the `/components` directory?

aduth replied:
Yes, maybe not `SkipToSelectedBlock` as-is, since it an implementation specific to block usage, but maybe a `SkipTo`, which `SkipToSelectedBlock` can extend (or just inline `SkipTo` where we're using `SkipToSelectedBlock`).

afercia replied:
Ah OK. Hm at the moment I don't see the need for other "Skip To"s, the ones in core and this one should suffice, for now. We can always iterate later, if need be?

aduth commented on editor/components/skip-to-selected-block/index.js:
```diff
@@ -0,0 +1,37 @@
+
+export default withSelect( ( select ) => {
+	return {
+		selectedBlock: select( 'core/editor' ).getSelectedBlock(),
```
`getSelectedBlock` can be a computationally-intensive selector. I wish we had a simpler way to get access to the direct UID value of the singular selected block. We do have `getBlockSelectionStart`, but this also returns a value for a multi-selection. Which actually raises another point: do we care if it's a singular block selection? Seems like we'd want to return to the first UID of the set of selected blocks in either case.

aduth commented on edit-post/assets/stylesheets/_z-index.scss:
```diff
@@ -55,6 +55,8 @@ $z-layers: (
 
 	'.components-autocomplete__results': 1000000,
+
+	'.skip-to-selected-block': 100000,
```
Why so high? In fact, even without `z-index` at all, it appears to layer as I'd expect.

FWIW, this isn't the same as putting the button on equal footing as `.edit-post-sidebar`, since it's already within the stacking context of the sidebar.

afercia replied:
For re-use in other places e.g. header toolbar, publishing flow, etc. I guess it should have a high z-index.

aduth commented on editor/components/skip-to-selected-block/index.js:
```diff
@@ -0,0 +1,37 @@
+	};
+
+	return (
+		selectedBlock && uid &&
```
`selectedBlock` will never be falsey here. You already have a condition above for `! selectedBlock`.

afercia replied:
`selectedBlock` is falsey initially and switching (even with mouse) the sidebar to the block inspector renders the skip link (the render method must return "something")

afercia commented:
> Is it usual to have such skips in different places on screens? It totally may be, which is why asking for insights.

I don't have an answer, I think this is the first time I see a "skip link" to actually jump back :) Will ask the a11y team opinion.

rianrietveld commented:
What's in a word: skip link / jump back. It's all relative to how some sees/uses the page I guess.
Keeping the wording uniform may be the clearest for users.
So "Skip to selected block" seems ok to me. Nice work Andrea :-)

afercia committed:
Proper CSS.

afercia committed:
Rename and document getBlockWrapperDOMNode.

afercia committed:
Use getBlockSelectionStart which also allows to simplify.

afercia commented:
I think I've addressed most of the points, pending feedback on https://github.com/WordPress/gutenberg/pull/5856#discussion_r178913504

For now, it's placed only in the block inspector. I've tested it also with multi-selection and thanks to `getBlockSelectionStart()` it would work fine. Worth noting the actual returned UID is the one of the block that gets selected first. For example, when multi-selecting starting from the last block and moving the mouse up, the UID is the one of the last block:

<img width="984" alt="screen shot 2018-04-04 at 17 35 30" src="https://user-images.githubusercontent.com/1682452/38318239-2b66c048-382f-11e8-8de8-119d6dd7c42f.png">

aduth reviewed (approved):
👍

afercia commented:
Rebased to solve conflict.

mtias commented:
Nice one!
